symengine/llvm_double.h

```diff
@@ -0,0 +1,432 @@
+        // will
+        // automatically append instructions to the basic block `BB'.
+        builder = llvm::make_unique<llvm::IRBuilder<>>(BB);
+        builder->SetInsertPoint(BB);
+        auto fmf = llvm::FastMathFlags();
+        // fmf.setUnsafeAlgebra();
+        builder->setFastMathFlags(fmf);
+
+        // Load all the symbols and create references
+        auto input_arg = &(*(F->args().begin()));
+        for (unsigned i = 0; i < inputs.size(); i++) {
+            if (not is_a<Symbol>(*inputs[i])) {
+                throw SymEngineException("Input contains a non-symbol.");
+            }
+            auto index
+                = llvm::ConstantInt::get(llvm::Type::getInt32Ty(*context), i);
+            auto ptr = builder->CreateGEP(llvm::Type::getDoubleTy(*context),
+                                          input_arg, index);
+            result_
+                = builder->CreateLoad(llvm::Type::getDoubleTy(*context), ptr);
+            symbol_ptrs.push_back(result_);
+        }
+
+        auto it = ++(F->args().begin());
+        auto out = &(*it);
+        std::vector<llvm::Value *> output_vals;
+
+        // Generate Ir for all the output exprs and save references
+        for (unsigned i = 0; i < outputs.size(); i++) {
+            output_vals.push_back(apply(*outputs[i]));
+        }
+
+        // Store all the output exprs at the end
+        for (unsigned i = 0; i < outputs.size(); i++) {
+            auto index
+                = llvm::ConstantInt::get(llvm::Type::getInt32Ty(*context), i);
+            auto ptr = builder->CreateGEP(llvm::Type::getDoubleTy(*context),
+                                          out, index);
+            builder->CreateStore(output_vals[i], ptr);
+        }
+
+        // Create the return instruction and add it to the basic block
+        builder->CreateRetVoid();
+
+        // Validate the generated code, checking for consistency.
+        llvm::verifyFunction(*F);
+
+        // std::cout << "LLVM IR" << std::endl;
+        // module->dump();
+
+        // Optimize the function.
+        fpm->run(*F);
+
+        // std::cout << "Optimized LLVM IR" << std::endl;
+        // module->dump();
+
+        // Now we create the JIT.
+        std::string error;
+        auto executionengine
+            = llvm::EngineBuilder(std::move(module))
+                  .setEngineKind(llvm::EngineKind::Kind::JIT)
+                  .setOptLevel(llvm::CodeGenOpt::Level::Aggressive)
+                  .setErrorStr(&error)
+                  .create();
+        // std::cout << error << std::endl;
+        executionengine->finalizeObject();
+
+        // Get the symbol's address
+        func = (intptr_t)executionengine->getPointerToFunction(F);
+    }
+
+    double call(const std::vector<double> &vec)
+    {
+        double ret;
+        ((double (*)(const double *, double *))func)(vec.data(), &ret);
+        return ret;
+    }
+
+    void call(double *outs, const double *inps)
```

Comment:
I don't if this has an actual effect, but theoretically it should help the compiler with optimizations:
void call(double * const outs, const double * const __restrict__ inps)